File: app/src/main/java/ua/meugen/android/pdfcomposer/app/PdfComposer.java
package ua.meugen.android.pdfcomposer.app;

import android.app.Application;
import android.arch.persistence.room.Room;
import android.content.Context;

import ua.meugen.android.pdfcomposer.app.di.AppComponent;
import ua.meugen.android.pdfcomposer.model.actions.AppActionApi;
import ua.meugen.android.pdfcomposer.model.actions.PdfExportActionApi;
import ua.meugen.android.pdfcomposer.model.actions.PdfExportRequest;
import ua.meugen.android.pdfcomposer.model.db.AppDatabase;
import ua.meugen.android.pdfcomposer.model.db.dao.PdfItemDao;
import ua.meugen.android.pdfcomposer.model.events.PdfExportProgressEvent;

public class PdfComposer extends Application implements AppComponent {

    public static AppComponent appComponent(final Context context) {
        return (AppComponent) context.getApplicationContext();
    }

    private PdfItemDao pdfItemDao;

    private AppDatabase createAppDatabase() {
        return Room
                .databaseBuilder(this, AppDatabase.class, "pdfcomposer")
                .fallbackToDestructiveMigration()
                .build();
    }

    @Override
    public PdfItemDao getPdfItemDao() {
        if (pdfItemDao == null) {
            pdfItemDao = createAppDatabase().pdfItemDao();
        }
        return pdfItemDao;
    }

    @Override
    public AppActionApi<PdfExportRequest, PdfExportProgressEvent> getPdfExportActionApi() {
        return new PdfExportActionApi(this, getPdfItemDao());
    }
}

File: app/src/main/java/ua/meugen/android/pdfcomposer/app/PdfComposer.kt
package ua.meugen.android.pdfcomposer.app

import android.app.Application
import android.content.Context

import ua.meugen.android.pdfcomposer.app.di.AppComponent
import ua.meugen.android.pdfcomposer.app.di.AppComponentImpl

class PdfComposer : Application() {

    lateinit var appComponent: AppComponent
        private set

    override fun onCreate() {
        super.onCreate()
        appComponent = AppComponentImpl(this)
    }

    companion object {

        fun from(context: Context): PdfComposer {
            return context.applicationContext as PdfComposer
        }
    }
}
